# Food Desert Investigation -- Data Wrangling and Cleaning

## Imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

I am going to use Pandas DataFrames to work with the csv files I've downloaded.

### State and County Data

I am filtering the StateAndCountyData to save the columns from 2016. I am doing this because all of the food data is from 2016. I will utilize other datasets to supplement this data with demographic data.

In [2]:
file_food_atlas = 'StateAndCountyData.csv'
data_state_county = pd.read_csv(file_food_atlas)
data_state_county.head()

FIPS State   County          Variable_Code         Value
0  1001    AL  Autauga          LACCESS_POP10  18428.439690
1  1001    AL  Autauga          LACCESS_POP15  17496.693040
2  1001    AL  Autauga  PCH_LACCESS_POP_10_15     -5.056026
3  1001    AL  Autauga      PCT_LACCESS_POP10     33.769657
4  1001    AL  Autauga      PCT_LACCESS_POP15     32.062255

In [3]:
filtered_df = data_state_county[data_state_county['Variable_Code'].str.contains('(16)$')]

/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/3569536789.py:1: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  filtered_df = data_state_county[data_state_county['Variable_Code'].str.contains('(16)$')]


In [4]:
filtered_df.head()

FIPS State   County      Variable_Code      Value
128276  1001    AL  Autauga             GROC16   3.000000
128277  1001    AL  Autauga     PCH_GROC_11_16 -40.000000
128279  1001    AL  Autauga          GROCPTH16   0.054271
128280  1001    AL  Autauga  PCH_GROCPTH_11_16 -40.085748
128282  1001    AL  Autauga           SUPERC16   1.000000

In [5]:
filtered_df['County'] = filtered_df['County'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
filtered_df['County'] = filtered_df['County'].str.replace('County', '') # Remove 'County' word, only present sometimes
filtered_df['County'] = filtered_df['County'].str.strip() # Remove leading and trailing spaces
filtered_df['County'] = filtered_df['County'].str.lower() # Convert to lowercase

/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/3599340196.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['County'] = filtered_df['County'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/3599340196.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['County'] = filtered_df['County'].str.replace('County', '') # Remove 'County' word, only present sometimes
/var/folders/d7/365gnrk10q72fy

In [6]:
filtered_df['State'] = filtered_df.loc[:, 'State'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
filtered_df['State'] = filtered_df['State'].str.strip() # Remove leading and trailing spaces
filtered_df['State'] = filtered_df['State'].str.upper() # Convert to uppercase

/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/3646267976.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['State'] = filtered_df.loc[:, 'State'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/3646267976.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df['State'] = filtered_df['State'].str.strip() # Remove leading and trailing spaces
/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ip

In [7]:
pivot_df = filtered_df.pivot_table(index=['FIPS', 'State', 'County'], columns='Variable_Code', values='Value').reset_index()

In [8]:
pivot_df.head()

Variable_Code  FIPS State County  CACFP_PART_2016  CONVS16  CONVSPTH16  FFR16  \
0                 1    AL  total         59447.50      NaN         NaN    NaN   
1                 2    AK  total         10039.00      NaN         NaN    NaN   
2                 4    AZ  total         42618.00      NaN         NaN    NaN   
3                 5    AR  total         66903.00      NaN         NaN    NaN   
4                 6    CA  total        595088.25      NaN         NaN    NaN   

Variable_Code  FFRPTH16  FSR16  FSRPTH16  ...  SNAP_PART_RATE16  \
0                   NaN    NaN       NaN  ...               NaN   
1                   NaN    NaN       NaN  ...               NaN   
2                   NaN    NaN       NaN  ...               NaN   
3                   NaN    NaN       NaN  ...               NaN   
4                   NaN    NaN       NaN  ...               NaN   

Variable_Code  SNAP_REPORTSIMPLE16  SPECS16  SPECSPTH16  SUPERC16  \
0                              NaN      NaN         NaN       NaN   
1                              NaN      NaN         NaN       NaN   
2                              NaN      NaN         NaN       NaN   
3                              NaN      NaN         NaN       NaN   
4                              NaN      NaN         NaN       NaN   

Variable_Code  SUPERCPTH16  State_Population_2016  WICS16  WICSPTH16  \
0                      NaN              4864745.0     NaN        NaN   
1                      NaN               741504.0     NaN        NaN   
2                      NaN              6945452.0     NaN        NaN   
3                      NaN              2990410.0     NaN        NaN   
4                      NaN             39209127.0     NaN        NaN   

Variable_Code  WIC_PART_2016  
0               1.291599e+05  
1               1.912092e+04  
2               1.639978e+05  
3               8.055467e+04  
4               1.174875e+06  

[5 rows x 55 columns]

### County Data

In [9]:
# Importing County Dataset
file_county_data = 'SupplementalDataCounty.csv'
data_county = pd.read_csv(file_county_data)
data_county.head()

FIPS State          County             Variable_Code  Value
0  1001    AL  Autauga County    2010_Census_Population  54571
1  1001    AL  Autauga County  Population_Estimate_2011  55208
2  1001    AL  Autauga County  Population_Estimate_2012  54936
3  1001    AL  Autauga County  Population_Estimate_2013  54713
4  1001    AL  Autauga County  Population_Estimate_2014  54876

In [10]:
filtered_counties = data_county[data_county['Variable_Code'].str.contains('(16)$')]

filtered_counties['County'] = filtered_counties['County'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
filtered_counties['County'] = filtered_counties['County'].str.replace('County', '') # Remove 'County' word, only present sometimes
filtered_counties['County'] = filtered_counties['County'].str.strip() # Remove leading and trailing spaces
filtered_counties['County'] = filtered_counties['County'].str.lower() # Convert to lowercase

filtered_counties['State'] = filtered_counties['State'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
filtered_counties['State'] = filtered_counties['State'].str.strip() # Remove leading and trailing spaces
filtered_counties['State'] = filtered_counties['State'].str.upper() # Convert to lowercase

counties = filtered_counties.pivot_table(index=['FIPS', 'State', 'County'], columns='Variable_Code', values='Value').reset_index()
counties.head()

/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/208533991.py:1: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  filtered_counties = data_county[data_county['Variable_Code'].str.contains('(16)$')]
/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/208533991.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_counties['County'] = filtered_counties['County'].str.replace(r'[^\w\s]', '', regex=True) # Remove punctuation
/var/folders/d7/365gnrk10q72fy9l9_q_hft80000gn/T/ipykernel_88901/208533991.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexe

Variable_Code  FIPS State   County  Population_Estimate_2016
0              1001    AL  autauga                     55242
1              1003    AL  baldwin                    207712
2              1005    AL  barbour                     25819
3              1007    AL     bibb                     22576
4              1009    AL   blount                     57517

In [11]:
counties = counties.drop(['State','County'], axis=1)

In [12]:
counties.head()

Variable_Code  FIPS  Population_Estimate_2016
0              1001                     55242
1              1003                    207712
2              1005                     25819
3              1007                     22576
4              1009                     57517

## Exploring pivot_df

In [13]:
pivot_df.head()

Variable_Code  FIPS State County  CACFP_PART_2016  CONVS16  CONVSPTH16  FFR16  \
0                 1    AL  total         59447.50      NaN         NaN    NaN   
1                 2    AK  total         10039.00      NaN         NaN    NaN   
2                 4    AZ  total         42618.00      NaN         NaN    NaN   
3                 5    AR  total         66903.00      NaN         NaN    NaN   
4                 6    CA  total        595088.25      NaN         NaN    NaN   

Variable_Code  FFRPTH16  FSR16  FSRPTH16  ...  SNAP_PART_RATE16  \
0                   NaN    NaN       NaN  ...               NaN   
1                   NaN    NaN       NaN  ...               NaN   
2                   NaN    NaN       NaN  ...               NaN   
3                   NaN    NaN       NaN  ...               NaN   
4                   NaN    NaN       NaN  ...               NaN   

Variable_Code  SNAP_REPORTSIMPLE16  SPECS16  SPECSPTH16  SUPERC16  \
0                              NaN      NaN         NaN       NaN   
1                              NaN      NaN         NaN       NaN   
2                              NaN      NaN         NaN       NaN   
3                              NaN      NaN         NaN       NaN   
4                              NaN      NaN         NaN       NaN   

Variable_Code  SUPERCPTH16  State_Population_2016  WICS16  WICSPTH16  \
0                      NaN              4864745.0     NaN        NaN   
1                      NaN               741504.0     NaN        NaN   
2                      NaN              6945452.0     NaN        NaN   
3                      NaN              2990410.0     NaN        NaN   
4                      NaN             39209127.0     NaN        NaN   

Variable_Code  WIC_PART_2016  
0               1.291599e+05  
1               1.912092e+04  
2               1.639978e+05  
3               8.055467e+04  
4               1.174875e+06  

[5 rows x 55 columns]

In [14]:
pivot_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3328 entries, 0 to 3327
Data columns (total 55 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   FIPS                      3328 non-null   int64  
 1   State                     3328 non-null   object 
 2   County                    3328 non-null   object 
 3   CACFP_PART_2016           51 non-null     float64
 4   CONVS16                   3143 non-null   float64
 5   CONVSPTH16                3143 non-null   float64
 6   FFR16                     3143 non-null   float64
 7   FFRPTH16                  3143 non-null   float64
 8   FSR16                     3143 non-null   float64
 9   FSRPTH16                  3143 non-null   float64
 10  GROC16                    3143 non-null   float64
 11  GROCPTH16                 3143 non-null   float64
 12  NSLP_PART_2016            51 non-null     float64
 13  PCH_CONVSPTH_11_16        3122 non-null   float64
 14  PCH_CONV

In [15]:
pivot_df.describe()

Variable_Code          FIPS  CACFP_PART_2016     CONVS16   CONVSPTH16  \
count           3328.000000        51.000000  3143.00000  3143.000000   
mean           29764.681490     85921.333333    40.35762     0.595571   
std            15635.421817    118749.706301    97.01666     0.304968   
min                1.000000      7532.750000     0.00000     0.000000   
25%            18110.500000     19351.875000     7.00000     0.401678   
50%            29072.000000     49185.750000    15.00000     0.542005   
75%            45067.500000     90559.125000    35.00000     0.722071   
max            56045.000000    595088.250000  2199.00000     3.680982   

Variable_Code        FFR16     FFRPTH16        FSR16     FSRPTH16  \
count          3143.000000  3143.000000  3143.000000  3143.000000   
mean             75.696787     0.584822    78.551066     0.775831   
std             268.845823     0.307703   269.993476     0.567851   
min               0.000000     0.000000     0.000000     0.000000   
25%               5.000000     0.418769     7.000000     0.495050   
50%              15.000000     0.588565    18.000000     0.672495   
75%              45.000000     0.748136    48.000000     0.894307   
max            8264.000000     5.805515  8113.000000    11.611030   

Variable_Code       GROC16    GROCPTH16  ...  SNAP_PART_RATE16  \
count          3143.000000  3143.000000  ...       3143.000000   
mean             20.807827     0.243835  ...         84.001372   
std              90.196535     0.216536  ...          9.223259   
min               0.000000     0.000000  ...         56.037000   
25%               2.000000     0.132983  ...         76.151000   
50%               5.000000     0.186116  ...         84.002000   
75%              12.000000     0.279887  ...         90.379000   
max            2495.000000     3.267974  ...        100.000000   

Variable_Code  SNAP_REPORTSIMPLE16      SPECS16   SPECSPTH16     SUPERC16  \
count                       3143.0  3143.000000  3143.000000  3143.000000   
mean                           1.0     7.458479     0.055450     1.782055   
std                            0.0    29.697479     0.075095     4.310650   
min                            1.0     0.000000     0.000000     0.000000   
25%                            1.0     0.000000     0.000000     0.000000   
50%                            1.0     1.000000     0.042268     1.000000   
75%                            1.0     4.000000     0.079363     1.500000   
max                            1.0   787.000000     1.095690    97.000000   

Variable_Code  SUPERCPTH16  State_Population_2016       WICS16    WICSPTH16  \
count          3143.000000           5.100000e+01  2985.000000  2982.000000   
mean              0.018255           6.334732e+06    14.577889     0.217240   
std               0.022460           7.242171e+06    48.084899     0.191315   
min               0.000000           5.842900e+05     1.000000     0.004938   
25%               0.000000           1.756930e+06     2.000000     0.114667   
50%               0.014654           4.438229e+06     5.000000     0.165353   
75%               0.029268           7.120066e+06    10.000000     0.247699   
max               0.250752           3.920913e+07  1188.000000     3.262643   

Variable_Code  WIC_PART_2016  
count           5.100000e+01  
mean            1.474524e+05  
std             2.065493e+05  
min             1.078283e+04  
25%             3.856721e+04  
50%             1.001519e+05  
75%             1.515955e+05  
max             1.174875e+06  

[8 rows x 53 columns]

In [16]:
pivot_df.columns

Index(['FIPS', 'State', 'County', 'CACFP_PART_2016', 'CONVS16', 'CONVSPTH16',
       'FFR16', 'FFRPTH16', 'FSR16', 'FSRPTH16', 'GROC16', 'GROCPTH16',
       'NSLP_PART_2016', 'PCH_CONVSPTH_11_16', 'PCH_CONVS_11_16',
       'PCH_FFRPTH_11_16', 'PCH_FFR_11_16', 'PCH_FSRPTH_11_16',
       'PCH_FSR_11_16', 'PCH_GROCPTH_11_16', 'PCH_GROC_11_16',
       'PCH_PC_WIC_REDEMP_11_16', 'PCH_RECFACPTH_11_16', 'PCH_RECFAC_11_16',
       'PCH_REDEMP_WICS_11_16', 'PCH_SPECSPTH_11_16', 'PCH_SPECS_11_16',
       'PCH_SUPERCPTH_11_16', 'PCH_SUPERC_11_16', 'PCH_WICINFANTCHILD_14_16',
       'PCH_WICSPTH_11_16', 'PCH_WICS_11_16', 'PCH_WICWOMEN_14_16',
       'PCT_WICINFANTCHILD16', 'PCT_WICWOMEN16', 'PC_WIC_REDEMP16',
       'Population_Estimate_2016', 'RECFAC16', 'RECFACPTH16', 'REDEMP_WICS16',
       'SBP_PART_2016', 'SFSP_PART_2016', 'SNAP_BBCE16', 'SNAP_CAP16',
       'SNAP_OAPP16', 'SNAP_PART_RATE16', 'SNAP_REPORTSIMPLE16', 'SPECS16',
       'SPECSPTH16', 'SUPERC16', 'SUPERCPTH16', 'State_Population_2

In [17]:
pivot_df.dtypes

Variable_Code
FIPS                          int64
State                        object
County                       object
CACFP_PART_2016             float64
CONVS16                     float64
CONVSPTH16                  float64
FFR16                       float64
FFRPTH16                    float64
FSR16                       float64
FSRPTH16                    float64
GROC16                      float64
GROCPTH16                   float64
NSLP_PART_2016              float64
PCH_CONVSPTH_11_16          float64
PCH_CONVS_11_16             float64
PCH_FFRPTH_11_16            float64
PCH_FFR_11_16               float64
PCH_FSRPTH_11_16            float64
PCH_FSR_11_16               float64
PCH_GROCPTH_11_16           float64
PCH_GROC_11_16              float64
PCH_PC_WIC_REDEMP_11_16     float64
PCH_RECFACPTH_11_16         float64
PCH_RECFAC_11_16            float64
PCH_REDEMP_WICS_11_16       float64
PCH_SPECSPTH_11_16          float64
PCH_SPECS_11_16             float64
PCH_SUPERCPTH_

## Cleaning

### Null values

In [18]:
pivot_df.isna()

Variable_Code   FIPS  State  County  CACFP_PART_2016  CONVS16  CONVSPTH16  \
0              False  False   False            False     True        True   
1              False  False   False            False     True        True   
2              False  False   False            False     True        True   
3              False  False   False            False     True        True   
4              False  False   False            False     True        True   
...              ...    ...     ...              ...      ...         ...   
3323           False  False   False             True    False       False   
3324           False  False   False             True    False       False   
3325           False  False   False             True    False       False   
3326           False  False   False             True    False       False   
3327           False  False   False             True    False       False   

Variable_Code  FFR16  FFRPTH16  FSR16  FSRPTH16  ...  SNAP_PART_RATE16  \
0               True      True   True      True  ...              True   
1               True      True   True      True  ...              True   
2               True      True   True      True  ...              True   
3               True      True   True      True  ...              True   
4               True      True   True      True  ...              True   
...              ...       ...    ...       ...  ...               ...   
3323           False     False  False     False  ...             False   
3324           False     False  False     False  ...             False   
3325           False     False  False     False  ...             False   
3326           False     False  False     False  ...             False   
3327           False     False  False     False  ...             False   

Variable_Code  SNAP_REPORTSIMPLE16  SPECS16  SPECSPTH16  SUPERC16  \
0                             True     True        True      True   
1                             True     True        True      True   
2                             True     True        True      True   
3                             True     True        True      True   
4                             True     True        True      True   
...                            ...      ...         ...       ...   
3323                         False    False       False     False   
3324                         False    False       False     False   
3325                         False    False       False     False   
3326                         False    False       False     False   
3327                         False    False       False     False   

Variable_Code  SUPERCPTH16  State_Population_2016  WICS16  WICSPTH16  \
0                     True                  False    True       True   
1                     True                  False    True       True   
2                     True                  False    True       True   
3                     True                  False    True       True   
4                     True                  False    True       True   
...                    ...                    ...     ...        ...   
3323                 False                   True   False      False   
3324                 False                   True   False      False   
3325                 False                   True   False      False   
3326                 False                   True   False      False   
3327                 False                   True   False      False   

Variable_Code  WIC_PART_2016  
0                      False  
1                      False  
2                      False  
3                      False  
4                      False  
...                      ...  
3323                    True  
3324                    True  
3325                    True  
3326                    True  
3327                    True  

[3328 rows x 55 columns]

In [19]:
pivot_df.isna().sum().sort_values(ascending=False)

Variable_Code
WIC_PART_2016               3277
SBP_PART_2016               3277
NSLP_PART_2016              3277
CACFP_PART_2016             3277
State_Population_2016       3277
SFSP_PART_2016              3277
PCH_PC_WIC_REDEMP_11_16     1520
PC_WIC_REDEMP16             1471
PCH_SPECS_11_16              435
PCH_SPECSPTH_11_16           432
PCH_WICSPTH_11_16            347
WICSPTH16                    346
PCH_REDEMP_WICS_11_16        345
PCH_WICS_11_16               344
REDEMP_WICS16                343
WICS16                       343
PCH_RECFAC_11_16             328
PCH_RECFACPTH_11_16          328
PCH_SUPERC_11_16             318
PCH_SUPERCPTH_11_16          317
PCH_FFR_11_16                242
PCH_FFRPTH_11_16             237
PCH_FSR_11_16                213
PCH_CONVS_11_16              211
PCH_FSRPTH_11_16             208
PCH_GROC_11_16               206
PCH_CONVSPTH_11_16           206
PCH_GROCPTH_11_16            201
Population_Estimate_2016     186
SNAP_REPORTSIMPLE16          

In [20]:
missing_sum = pivot_df.isnull().sum()
missing_percent = 100 * pivot_df.isnull().mean()
missing = pd.concat([missing_sum, missing_percent], axis=1)
missing = missing.sort_values([0, 1], ascending=False)

missing = missing.rename(columns = {1:'Percent'})
missing = missing.rename(columns = {0:'Sum'})

missing[missing.Percent > 50]

Sum    Percent
Variable_Code                         
CACFP_PART_2016        3277  98.467548
NSLP_PART_2016         3277  98.467548
SBP_PART_2016          3277  98.467548
SFSP_PART_2016         3277  98.467548
State_Population_2016  3277  98.467548
WIC_PART_2016          3277  98.467548

The above columns ('CACFP_PART_2016', 'NSLP_PART_2016', 'SBP_PART_2016', 'SFSP_PART_2016', 'State_Population_2016', 'WIC_PART_2016') are all state totals. Instead of dropping these columns, I am going to input the state total for every county in the state. This will allow me to keep these columns as part of the data.

In [21]:
def fill_missing(df, col):
    state_total = df.groupby("State")[col].sum().to_dict()
    df[col] = df[col].fillna(df["State"].map(state_total))
    return df[col]

fill_missing(pivot_df, 'CACFP_PART_2016')

0        59447.50
1        10039.00
2        42618.00
3        66903.00
4       595088.25
          ...    
3323      7532.75
3324      7532.75
3325      7532.75
3326      7532.75
3327      7532.75
Name: CACFP_PART_2016, Length: 3328, dtype: float64

In [22]:
# Checking my method. It works!
pivot_df[pivot_df['State']=='AK']['CACFP_PART_2016']

1      10039.0
118    10039.0
119    10039.0
120    10039.0
121    10039.0
122    10039.0
123    10039.0
124    10039.0
125    10039.0
126    10039.0
127    10039.0
128    10039.0
129    10039.0
130    10039.0
131    10039.0
132    10039.0
133    10039.0
134    10039.0
135    10039.0
136    10039.0
137    10039.0
138    10039.0
139    10039.0
140    10039.0
141    10039.0
142    10039.0
143    10039.0
144    10039.0
145    10039.0
146    10039.0
147    10039.0
148    10039.0
149    10039.0
150    10039.0
151    10039.0
152    10039.0
153    10039.0
154    10039.0
155    10039.0
156    10039.0
157    10039.0
158    10039.0
159    10039.0
160    10039.0
161    10039.0
162    10039.0
163    10039.0
164    10039.0
165    10039.0
166    10039.0
167    10039.0
168    10039.0
169    10039.0
170    10039.0
171    10039.0
172    10039.0
173    10039.0
174    10039.0
175    10039.0
Name: CACFP_PART_2016, dtype: float64

In [23]:
fill_missing(pivot_df, 'NSLP_PART_2016')
fill_missing(pivot_df, 'SBP_PART_2016')
fill_missing(pivot_df, 'SFSP_PART_2016')
fill_missing(pivot_df, 'State_Population_2016')
fill_missing(pivot_df, 'WIC_PART_2016')

0       1.291599e+05
1       1.912092e+04
2       1.639978e+05
3       8.055467e+04
4       1.174875e+06
            ...     
3323    1.078283e+04
3324    1.078283e+04
3325    1.078283e+04
3326    1.078283e+04
3327    1.078283e+04
Name: WIC_PART_2016, Length: 3328, dtype: float64

In [24]:
pivot_df.isna().sum().sort_values(ascending=False)

Variable_Code
PCH_PC_WIC_REDEMP_11_16     1520
PC_WIC_REDEMP16             1471
PCH_SPECS_11_16              435
PCH_SPECSPTH_11_16           432
PCH_WICSPTH_11_16            347
WICSPTH16                    346
PCH_REDEMP_WICS_11_16        345
PCH_WICS_11_16               344
WICS16                       343
REDEMP_WICS16                343
PCH_RECFAC_11_16             328
PCH_RECFACPTH_11_16          328
PCH_SUPERC_11_16             318
PCH_SUPERCPTH_11_16          317
PCH_FFR_11_16                242
PCH_FFRPTH_11_16             237
PCH_FSR_11_16                213
PCH_CONVS_11_16              211
PCH_FSRPTH_11_16             208
PCH_GROC_11_16               206
PCH_CONVSPTH_11_16           206
PCH_GROCPTH_11_16            201
Population_Estimate_2016     186
SUPERC16                     185
RECFAC16                     185
RECFACPTH16                  185
CONVS16                      185
SNAP_CAP16                   185
SPECSPTH16                   185
SNAP_BBCE16                  

PCH_PC_WIC_REDEMP_11_16 is WIC redemptions per capita (% change), 2011-2016. Since I am not including any data from 2011, I am going to drop this column. It also happens to be a column with many missing values.

In [25]:
pivot_df = pivot_df.drop('PCH_PC_WIC_REDEMP_11_16', axis=1)

As with the column above, I will drop every column that contains a year other than 2016.

In [26]:
pivot_df = pivot_df[pivot_df.columns.drop(list(pivot_df.filter(regex='11|14')))]
pivot_df.columns

Index(['FIPS', 'State', 'County', 'CACFP_PART_2016', 'CONVS16', 'CONVSPTH16',
       'FFR16', 'FFRPTH16', 'FSR16', 'FSRPTH16', 'GROC16', 'GROCPTH16',
       'NSLP_PART_2016', 'PCT_WICINFANTCHILD16', 'PCT_WICWOMEN16',
       'PC_WIC_REDEMP16', 'Population_Estimate_2016', 'RECFAC16',
       'RECFACPTH16', 'REDEMP_WICS16', 'SBP_PART_2016', 'SFSP_PART_2016',
       'SNAP_BBCE16', 'SNAP_CAP16', 'SNAP_OAPP16', 'SNAP_PART_RATE16',
       'SNAP_REPORTSIMPLE16', 'SPECS16', 'SPECSPTH16', 'SUPERC16',
       'SUPERCPTH16', 'State_Population_2016', 'WICS16', 'WICSPTH16',
       'WIC_PART_2016'],
      dtype='object', name='Variable_Code')

Checking for null values after removing columns containing dates other than 2016.

In [27]:
pivot_df.isna().sum().sort_values(ascending=False)

Variable_Code
PC_WIC_REDEMP16             1471
WICSPTH16                    346
WICS16                       343
REDEMP_WICS16                343
Population_Estimate_2016     186
RECFAC16                     185
SNAP_PART_RATE16             185
SNAP_OAPP16                  185
SNAP_CAP16                   185
SNAP_BBCE16                  185
SPECSPTH16                   185
SUPERC16                     185
RECFACPTH16                  185
SUPERCPTH16                  185
PCT_WICINFANTCHILD16         185
PCT_WICWOMEN16               185
SPECS16                      185
GROCPTH16                    185
GROC16                       185
FSRPTH16                     185
FSR16                        185
FFRPTH16                     185
FFR16                        185
CONVSPTH16                   185
CONVS16                      185
SNAP_REPORTSIMPLE16          185
State_Population_2016          0
FIPS                           0
SFSP_PART_2016                 0
SBP_PART_2016                

WIC is the Special Supplemental Nutrition Program for Women, Infants, and Children (WIC). For more information, please visit this link: https://www.fns.usda.gov/wic

I am dropping the column, PC_WIC_REDEMP16. This column is missing 1471 values, and I believe I can access information regarding WIC through other variables such as WICSPTH16 (WIC-authorized stores/1,000 pop), WICS16 (WIC-authorized stores), REDEMP_WICS16 (WIC redemptions/WIC-authorized stores), PCT_WICINFANTCHILD16 (WIC % infant and children participants), and PCT_WICWOMEN16 (WIC % women participants).

These columns are also missing quite a few values, but I will impute the median value by state for each missing value.

In [28]:
pivot_df = pivot_df.drop('PC_WIC_REDEMP16', axis=1)

In [29]:
def impute_median(df, col):
    state_total = df.groupby("State")[col].median().to_dict()
    df[col] = df[col].fillna(df["State"].map(state_total))
    return df[col]

In [30]:
impute_median(pivot_df, 'WICSPTH16')
impute_median(pivot_df, 'WICS16')
impute_median(pivot_df, 'REDEMP_WICS16')
impute_median(pivot_df, 'PCT_WICINFANTCHILD16')
impute_median(pivot_df, 'PCT_WICWOMEN16')

0       3.309759
1       3.065642
2       2.850142
3       3.595835
4       3.249196
          ...   
3323    2.375607
3324    2.375607
3325    2.375607
3326    2.375607
3327    2.375607
Name: PCT_WICWOMEN16, Length: 3328, dtype: float64

In [31]:
pivot_df.isna().sum().sort_values(ascending=True)

Variable_Code
FIPS                          0
WICS16                        0
State_Population_2016         0
SFSP_PART_2016                0
SBP_PART_2016                 0
REDEMP_WICS16                 0
WICSPTH16                     0
PCT_WICWOMEN16                0
PCT_WICINFANTCHILD16          0
NSLP_PART_2016                0
WIC_PART_2016                 0
State                         0
County                        0
CACFP_PART_2016               0
GROCPTH16                   185
SUPERCPTH16                 185
SUPERC16                    185
SPECSPTH16                  185
SPECS16                     185
SNAP_REPORTSIMPLE16         185
SNAP_PART_RATE16            185
SNAP_OAPP16                 185
SNAP_CAP16                  185
SNAP_BBCE16                 185
CONVS16                     185
CONVSPTH16                  185
RECFACPTH16                 185
FFR16                       185
FFRPTH16                    185
FSR16                       185
FSRPTH16                  

SNAP is the Supplemental Nutrition Assistance Program. SNAP offers benefits on EBT cards for low income families to supplement their grocery store budgets. Read more about SNAP here: https://www.fns.usda.gov/snap/supplemental-nutrition-assistance-program

I am going to impute the median value by state for each county row for the column, SNAP_PART_RATE16. This column represents SNAP participants (% eligible population).

In [32]:
impute_median(pivot_df, 'SNAP_PART_RATE16')

0       86.898
1       71.454
2       73.500
3       72.327
4       71.944
         ...  
3323    56.037
3324    56.037
3325    56.037
3326    56.037
3327    56.037
Name: SNAP_PART_RATE16, Length: 3328, dtype: float64

In [33]:
pivot_df[['GROC16']]

Variable_Code  GROC16
0                 NaN
1                 NaN
2                 NaN
3                 NaN
4                 NaN
...               ...
3323              4.0
3324             11.0
3325              2.0
3326              2.0
3327              4.0

[3328 rows x 1 columns]

In [34]:
pivot_df[pivot_df.index==0]

Variable_Code  FIPS State County  CACFP_PART_2016  CONVS16  CONVSPTH16  FFR16  \
0                 1    AL  total          59447.5      NaN         NaN    NaN   

Variable_Code  FFRPTH16  FSR16  FSRPTH16  ...  SNAP_PART_RATE16  \
0                   NaN    NaN       NaN  ...            86.898   

Variable_Code  SNAP_REPORTSIMPLE16  SPECS16  SPECSPTH16  SUPERC16  \
0                              NaN      NaN         NaN       NaN   

Variable_Code  SUPERCPTH16  State_Population_2016  WICS16  WICSPTH16  \
0                      NaN              4864745.0     6.0   0.178992   

Variable_Code  WIC_PART_2016  
0                129159.9167  

[1 rows x 34 columns]

In [35]:
pivot_df[pivot_df.index==1]

Variable_Code  FIPS State County  CACFP_PART_2016  CONVS16  CONVSPTH16  FFR16  \
1                 2    AK  total          10039.0      NaN         NaN    NaN   

Variable_Code  FFRPTH16  FSR16  FSRPTH16  ...  SNAP_PART_RATE16  \
1                   NaN    NaN       NaN  ...            71.454   

Variable_Code  SNAP_REPORTSIMPLE16  SPECS16  SPECSPTH16  SUPERC16  \
1                              NaN      NaN         NaN       NaN   

Variable_Code  SUPERCPTH16  State_Population_2016  WICS16  WICSPTH16  \
1                      NaN               741504.0     3.0   0.615642   

Variable_Code  WIC_PART_2016  
1                 19120.9167  

[1 rows x 34 columns]

In [36]:
pivot_df[pivot_df.index==3]

Variable_Code  FIPS State County  CACFP_PART_2016  CONVS16  CONVSPTH16  FFR16  \
3                 5    AR  total          66903.0      NaN         NaN    NaN   

Variable_Code  FFRPTH16  FSR16  FSRPTH16  ...  SNAP_PART_RATE16  \
3                   NaN    NaN       NaN  ...            72.327   

Variable_Code  SNAP_REPORTSIMPLE16  SPECS16  SPECSPTH16  SUPERC16  \
3                              NaN      NaN         NaN       NaN   

Variable_Code  SUPERCPTH16  State_Population_2016  WICS16  WICSPTH16  \
3                      NaN              2990410.0     4.0   0.147361   

Variable_Code  WIC_PART_2016  
3                 80554.6667  

[1 rows x 34 columns]

It looks like observations that are missing values for GROC16 are also missing many other column values; therefore, I am going to drop all observations with missing data for GROC16.

In [37]:
pivot_df = pivot_df[pivot_df['GROC16'].notna()]

In [38]:
pivot_df[['GROC16']]

Variable_Code  GROC16
51                3.0
52               29.0
53                4.0
54                5.0
55                5.0
...               ...
3323              4.0
3324             11.0
3325              2.0
3326              2.0
3327              4.0

[3143 rows x 1 columns]

In [39]:
pivot_df.isna().sum().sort_values(ascending=False)

Variable_Code
Population_Estimate_2016    135
FIPS                          0
SNAP_REPORTSIMPLE16           0
SFSP_PART_2016                0
SNAP_BBCE16                   0
SNAP_CAP16                    0
SNAP_OAPP16                   0
SNAP_PART_RATE16              0
SPECS16                       0
REDEMP_WICS16                 0
SPECSPTH16                    0
SUPERC16                      0
SUPERCPTH16                   0
State_Population_2016         0
WICS16                        0
WICSPTH16                     0
SBP_PART_2016                 0
RECFACPTH16                   0
State                         0
RECFAC16                      0
PCT_WICWOMEN16                0
PCT_WICINFANTCHILD16          0
NSLP_PART_2016                0
GROCPTH16                     0
GROC16                        0
FSRPTH16                      0
FSR16                         0
FFRPTH16                      0
FFR16                         0
CONVSPTH16                    0
CONVS16                   

The only column that still contains missing values is Population_Estimate_2016, so I will use the Population_Estimate_2016 column from the counties DataFrame.

In [40]:
pivot_df = pivot_df.drop('Population_Estimate_2016', axis=1)

In [41]:
pivot_df = pd.merge(pivot_df, counties, on='FIPS')

In [42]:
pivot_df.isna().sum()

Variable_Code
FIPS                        0
State                       0
County                      0
CACFP_PART_2016             0
CONVS16                     0
CONVSPTH16                  0
FFR16                       0
FFRPTH16                    0
FSR16                       0
FSRPTH16                    0
GROC16                      0
GROCPTH16                   0
NSLP_PART_2016              0
PCT_WICINFANTCHILD16        0
PCT_WICWOMEN16              0
RECFAC16                    0
RECFACPTH16                 0
REDEMP_WICS16               0
SBP_PART_2016               0
SFSP_PART_2016              0
SNAP_BBCE16                 0
SNAP_CAP16                  0
SNAP_OAPP16                 0
SNAP_PART_RATE16            0
SNAP_REPORTSIMPLE16         0
SPECS16                     0
SPECSPTH16                  0
SUPERC16                    0
SUPERCPTH16                 0
State_Population_2016       0
WICS16                      0
WICSPTH16                   0
WIC_PART_2016             

### Changing data types

In [43]:
pivot_df.dtypes

Variable_Code
FIPS                          int64
State                        object
County                       object
CACFP_PART_2016             float64
CONVS16                     float64
CONVSPTH16                  float64
FFR16                       float64
FFRPTH16                    float64
FSR16                       float64
FSRPTH16                    float64
GROC16                      float64
GROCPTH16                   float64
NSLP_PART_2016              float64
PCT_WICINFANTCHILD16        float64
PCT_WICWOMEN16              float64
RECFAC16                    float64
RECFACPTH16                 float64
REDEMP_WICS16               float64
SBP_PART_2016               float64
SFSP_PART_2016              float64
SNAP_BBCE16                 float64
SNAP_CAP16                  float64
SNAP_OAPP16                 float64
SNAP_PART_RATE16            float64
SNAP_REPORTSIMPLE16         float64
SPECS16                     float64
SPECSPTH16                  float64
SUPERC16      

#### State_Population_2016
This should not be a float variable because people cannot be counted as decimals.

In [44]:
pivot_df['State_Population_2016'] = pivot_df['State_Population_2016'].astype('int')

#### Population_Estimate_2016
This should not be a float variable for the same reason as State_Population_2016. People are counted in whole numbers.

In [45]:
pivot_df['Population_Estimate_2016'] = pivot_df['Population_Estimate_2016'].astype('int')

#### Grocery and Convenience stores
The columns listed below are all counts of physical locations, so they cannot be floats. They must be integers.

In [46]:
pivot_df['GROC16'] = pivot_df['GROC16'].astype('int') # Count of grocery stores
pivot_df['CONVS16'] = pivot_df['CONVS16'].astype('int') # Count of convenience stores
pivot_df['RECFAC16'] = pivot_df['RECFAC16'].astype('int') # Count of recreation & fitness facilities
pivot_df['SPECS16'] = pivot_df['SPECS16'].astype('int') # Count of specialized food stores
pivot_df['SUPERC16'] = pivot_df['SUPERC16'].astype('int') # Count of supercenters & club stores
pivot_df['FFR16'] = pivot_df['FFR16'].astype('int') # Count of fast-food restaurants
pivot_df['FSR16'] = pivot_df['FSR16'].astype('int') # Count of full-service restaurants

In [47]:
pivot_df.dtypes

Variable_Code
FIPS                          int64
State                        object
County                       object
CACFP_PART_2016             float64
CONVS16                       int64
CONVSPTH16                  float64
FFR16                         int64
FFRPTH16                    float64
FSR16                         int64
FSRPTH16                    float64
GROC16                        int64
GROCPTH16                   float64
NSLP_PART_2016              float64
PCT_WICINFANTCHILD16        float64
PCT_WICWOMEN16              float64
RECFAC16                      int64
RECFACPTH16                 float64
REDEMP_WICS16               float64
SBP_PART_2016               float64
SFSP_PART_2016              float64
SNAP_BBCE16                 float64
SNAP_CAP16                  float64
SNAP_OAPP16                 float64
SNAP_PART_RATE16            float64
SNAP_REPORTSIMPLE16         float64
SPECS16                       int64
SPECSPTH16                  float64
SUPERC16      

All columns pertaining to counts of stores per 1,000 people must also be integers.

In [48]:
pivot_df['RECFACPTH16'] = pivot_df['RECFACPTH16'].astype('int') # Count of recreation & fitness facilities per 1,000 pop
pivot_df['CONVSPTH16'] = pivot_df['CONVSPTH16'].astype('int') # Count of convenience stores per 1,000 pop
pivot_df['FFRPTH16'] = pivot_df['FFRPTH16'].astype('int') # Count of fast-food restaurants per 1,000 pop
pivot_df['FSRPTH16'] = pivot_df['FSRPTH16'].astype('int') # Count of full-service restaurants per 1,000 pop
pivot_df['GROCPTH16'] = pivot_df['GROCPTH16'].astype('int') # Count of grocery stores per 1,000 pop
pivot_df['SPECSPTH16'] = pivot_df['SPECSPTH16'].astype('int') # Count of specialized food stores per 1,000 pop
pivot_df['SUPERCPTH16'] = pivot_df['SUPERCPTH16'].astype('int') # Count of supercenter food & club stores per 1,000 pop

In [49]:
pivot_df.dtypes

Variable_Code
FIPS                          int64
State                        object
County                       object
CACFP_PART_2016             float64
CONVS16                       int64
CONVSPTH16                    int64
FFR16                         int64
FFRPTH16                      int64
FSR16                         int64
FSRPTH16                      int64
GROC16                        int64
GROCPTH16                     int64
NSLP_PART_2016              float64
PCT_WICINFANTCHILD16        float64
PCT_WICWOMEN16              float64
RECFAC16                      int64
RECFACPTH16                   int64
REDEMP_WICS16               float64
SBP_PART_2016               float64
SFSP_PART_2016              float64
SNAP_BBCE16                 float64
SNAP_CAP16                  float64
SNAP_OAPP16                 float64
SNAP_PART_RATE16            float64
SNAP_REPORTSIMPLE16         float64
SPECS16                       int64
SPECSPTH16                    int64
SUPERC16      

## Demographic Data